src/client/getPage.client.ts

```diff
@@ -40,17 +41,20 @@ async function getPageById(pageId: string): Promise<any> {
 function getPageInfo(): {
   pageId: string
   pageProps: Record<string, unknown>
+  routes: PageRoute[]
 } {
   const pageId = window.__vite_plugin_ssr.pageId
   const pageProps = window.__vite_plugin_ssr.pageProps
-  return { pageId, pageProps }
+  const routes = window.__vite_plugin_ssr.routes
```

Comment:
I don't think we need this; `loadPageRoutes()` provides all the routing information of all pages.

Reply:
Route functions will be returned as-is from loadPageRoutes. This is unusable in the client because the functions depend on contextProps, which are not available in client code.

Reply:
It also means that the routes can vary on a page-by-page basis. I would much prefer if we could simply load one set of routes for the entire application, but unfortunately that is not possible with the features you've exposed.

Reply:
How about forbidding Route Functions when using Vue Router? I'd rather have a simple MVP first then we can see whether we want more features.

If we remove Route Function support, we can get rid of `window.__vite_plugin_ssr.routes` and use `loadPageRoutes()` instead right?

Reply:
I simplified this bit after merging in your contextProps changes. Instead of separately handling routes, I just made it a built-in part of the context (similar to url) and pushed it into the `passToClient` array so the client will always have it in context. 

> If we remove Route Function support, we can get rid of window.__vite_plugin_ssr.routes and use loadPageRoutes() instead right?

I already removed `window.__vite_plugin_ssr.routes` as mentioned above. We do still need to keep a copy of the routes in the context for all rendered paths, because as I mentioned before there is a possibility the user is using page-specific context in the creation of that route. If you wanted to avoid this overhead, removing the route functions feature would allow us to use a single set of routes for the entire application. We would still need to get that set of routes to the client somehow, so if we're removing the routes from the contextProps, then we'd need another solution.

Reply:
> If you wanted to avoid this overhead, removing the route functions

Yes, that's what I have in mind. I'm thinking of doing some simplifications to Route Functions so I'd suggest to add the Route Function feature to Vue Router at a later point.

>  We would still need to get that set of routes to the client somehow

I believe `loadPageRoutes()` is enough. It loads all the route information we need (inlcuding all route strings and all route functions) on the client.

Reply:
> I believe loadPageRoutes() is enough. It loads all the route information we need (inlcuding all route strings and all route functions) on the client.

But currently the route functions need to be compiled on the server as they expect a full contextProps object. Are you proposing changing it so that only client-available context props are available to route functions? 

There is still the problem of the return value of these functions being something Vue Router doesn't know what to do with. Even if I am able to compile the route function in the client, there's not a whole lot I can do with `{ matchValue: number, routeParams: Object }` since Vue Router is the matcher. If the functions were to instead return dynamically generated paths, then we could recompute all the functions on navigation and update the routes. That is what I am currently supporting in the vue router plugin. If the route function gets a non-string, it will throw an error.

Reply:
What I'm suggesting is to completely forbid the usage of Route Functions when using the Vue Router plugin. I suggest to add this functionality later.

`loadPageRoutes()` should be sufficient then.

> only client-available context props are available to route functions?

FYI this is actually already the case. (That's actually a problem for some use cases and I've plans to solve this.)

Reply:
Ok but we will still need a strategy for getting the routes to the client. Currently they're in every contextProps payload. If we say we're dropping support for route functions when using the vue-router integration, we only have to get the routes once, but we still need to get them once. If we're removing the routes member from the contextProps, where would you like to store them? You want to call `loadPageRoutes` in the client? IMO this is unnecessary network overhead. Additionally, `loadPageRoutes` is not actually a comprehensive set of routes - it does not include the implicit ones that do not have a corresponding route file.

Reply:
> You want to call loadPageRoutes in the client?

This is already the case: the client loads `route.shared.ts` which calls `loadPageRoutes()`.

>  it does not include the implicit ones that do not have a corresponding route file.

The `route.shared.ts` also knows about all `.page.js` paths.

Also `loadPageRoutes()` works for Route Functions whereas  your serialization approach doesn't. (When using `vite-plugin-ssr` without Vue Router.)

Reply:
> This is already the case: the client loads route.shared.ts which calls loadPageRoutes().

It does? Which client? You mean the built-in client router? Remember, I'm working on an alternative to that code. In the implementations I've done I have no XHR traffic on load, and the only XHR traffic on navigation is for `index.pageProps.json` resources - exactly as it should be, IMO. Any network overhead on load, or additional requests being made on navigation would be wasteful.

Reply:
In production these requests are preloaded.

Reply:
It's still another round trip. With Core Web Vitals rollout coming soon, we need to be absolutely meticulous about client performance. We cannot introduce possible bottlenecks. This is a minimal amount of information that could easily be sent to the user as part of another payload.  I see no benefit to loading it as a separate resource in the client.

Reply:
One way to think about this is that route files are just like any other source code and are loaded the same way than code loaded with `<script>` tags. (Code that is preloaded with a preload tag and then used with `import.meta.glob` is, from a performance perspective, equivalent to a `<script>` tag.)

I'm all for performance improvements, but tackling the performance problems of `<script>` tags is out of scope for `vite-plugin-ssr`.

Injecting the route strings into HTML (this is basically the only thing you achieve with your approach; the info for filesystem routing is already injected in source code by Vite) solves the `<script>`-tag-delay problem for only 0.01% (probably even less) of source code.

Reply:
Maybe I'm not understanding your solution. I was under the impression that the route files existed outside the context of the bundle and thus were being loaded via AJAX (like the contextProps.json files). If that's not the case and you're telling me these files are are part of the Vite bundle, then I'm not as concerned. Thanks for clarifying.

So just to clarify on the solution you're proposing then:

- In the client Vue Router plugin, load route strings and functions via `loadPageRoutes` 
- For now, throw if we encounter a route function in the loaded routes.
- merge these routes with implicit routes (for pages with no route files). (I'm still not 100% clear on how you're saying we get these in the client. IMO it would be nice if the implicit routes were simply included with the explicit ones.)

Reply:
Sounds good.

> merge these routes with implicit routes (for pages with no route files). (I'm still not 100% clear on how you're saying we get these in the client. IMO it would be nice if the implicit routes were simply included with the explicit ones.)

The trick here is that we can determine the filesystem routes by using the list of page ids (the id of a page is the base path of its page files and we can use that to determe its filesystem route), see how `routeWith_filesystem` works.

>  part of the Vite bundle

Yes, check out `dist/server/pageFiles.node.js` to see how it works. You'll also see that the `.page.route.js` files are also in there.

Reply:
I'm having a problem implementing this solution. Your proposed method for overriding the matching behavior (passing it into the `PageRender` as a customRouting option) involves setting the overrides to the `ssrEnv`. The `ssrEnv` code seems to assume it is executed in a Node environment, but this makes `route.shared.ts` incompatible with browser environments if I reference anything from the `ssrEnv`. This manifests in the form of a `global is not defined` error when trying to call any function that references my `matchRoutes` and `sortRoutes` override functions.

How would you like to solve this? My solution before was to pre-compile the routes on the server and cache them for the client to pick up, but you're pushing back on that so I'm kind of out of ideas here.

Reply:
If you want to take a look at the problem, I've pushed the code in its broken state. You can try checking out my branch, npm installing at root, then starting the vue-router example. BTW I added the vue-router example as a package in the workspace so that it will get all the symlinks from the other packages.

Reply:
I wouldn't use `ssrEnv`, but instead implement a custom mechanism that, contrary to `ssrEnv`, is isomorphic.

For example, something along these lines.

```js
// Browser, user-land

// /pages/_default.page.client.js

import { useVueRouter } from '@vite-plugin-ssr/vue-router/client'

useVueRouter()
```

```js
// Browser, `@vite-plugin-ssr/vue-router/client`-land

import { setCustomRouter } from 'vite-plugin-ssr/client/router'

export { useVueRouter }

function useVueRouter() {
  setCustomRouter(/*...*/)
}
```

```js
// Browser, `vite-plugin-ssr/client/router`-land

export { setCustomRouter }

function setCustomRouter(customRouter) {
  window.customRouter = customRouter
}
```

```js
// Isomorphic, `route.shared.ts`-land

function getCustomRouter() {
    if( typeof window !== "undefined") {
      // Browser
      return window.customRouter
    } else {
      // Node.js; `global.customRouter` has been set somewhere during the `createPageRender()` call.
      return global.customRouter
    }
}
```

The Node.js story starting with `createPageRender({customRouter})` would be analogous.

Let me know if something's not clear.

Reply:
OK I'll give this a go

Reply:
This almost works, but I'm getting a strange error when trying to call `loadPageRoutes` on the server (seems to work OK in the client). 

```
Error: [vite-plugin-ssr@0.1.0-beta.34][Internal Failure] You stumbled upon an bug in `vite-plugin-ssr`'s source code (an internal assertion failed). This should definitely not be happening, and you should create a new issue at https://github.com/brillout/vite-plugin-ssr/issues/new that includes this error stack (the error stack is enough to debug this). Or reach out on Discord. A fix will be written promptly.
    at Module.getPageFiles (/REDACTED/code/git-repos/vite-plugin-ssr/src/page-files/getPageFiles.shared.ts:37:9)
    at Module.getPageIds (/REDACTED/code/git-repos/vite-plugin-ssr/src/route.shared.ts:172:43)
    at Module.getRoutes (/REDACTED/code/git-repos/vite-plugin-ssr/packages/vue-router/isomorphic/get-routes.ts:5:68)
    at initRoutes (/@fs/REDACTED/code/git-repos/vite-plugin-ssr/packages/vue-router/server/plugin.ts:22:52)
    at async eval (/REDACTED/code/git-repos/vite-plugin-ssr/packages/vue-router/server/plugin.ts:51:17)
```
Console logging indicates there are some cases where `allPageFiles` is not set when `getPageFiles` is called, I think causing the `assert(hasProp(allPageFiles, '.page'))` assertion to fail. I don't know enough about how this works to troubleshoot why this would ever be the case.

My latest changes are pushed up to this branch if you want to try debugging there.

Reply:
Add `@vite-plugin-ssr/vue-router` to Vite's `config.ssr.external`. You can add it directly at `src/plugin/dev.ts` and `src/plugin/build.ts` if you want.

Reply:
OK, doing that sort of worked but it sent me down a rabbit hole of debugging module resolution issues. My current understanding is that these files are now getting resolved here: https://github.com/vitejs/vite/blob/eb66b4350c635fb4f2ef2e8a9eb50958cde73743/packages/vite/src/node/ssr/ssrModuleLoader.ts#L93

However, as you can see, this is a direct call into the `resolve` package, which as indicated here, does not support the node package `exports` field: https://github.com/browserify/resolve/issues/222. More context here: https://github.com/vitejs/vite/issues/1418 it would appear there is SOME mitigation in place in Vite for this shortcoming of resolve, but perhaps something about these packages being marked external is bypassing Vite's handling of the `exports` field? 

All I know is I can get this working if I hack the compiled files of the vue-router package into the root of the package dir (so `resolve` finds the files it expects by virtue of the fact that they literally exist at the imported path). I've pushed it up in this state. It works, kinda, but it sucks. Not sure how you want to work around this.

The good news is that despite this module resolution nonsense, the routing functionality is working!

Reply:
For Node.js, we don't need Vite for any of our souce code: `vite-plugin-ssr` only makes user files (i.e. `**/*.page.*` files and their user-land dependencies) be processed by Vite, which is good thing as it simplifies things. (From Vite's perspective the Node.js entry point is `src/page-files/pageFiles.node.ts`.)

Yes, I had problems this as well and I'm not entirely sure how Vite loads dependencies, but what I know is that we don't need something as drastic as `cp -r ./dist/* ./`: instead, we can create some `.js` files at the right positions to do something like `src/client/router/index.js` does. When you think about it: once Vite does `require()`, then we are in Node.js land and Vite is out of the loop and we don't need to hack at this point anymore; we only need to hack the entry points between the user files and our source code.

Reply:
Got it, yeah that's a bit less crappy of a workaround isn't it haha. Went ahead and applied that to the vue-router package. I'm going to mark this thread resolved now

Reply:
🎉 

> yeah that's a bit less crappy of a workaround isn't it haha

:-)